patchset/0057-avfilter-vf_xcam-add-xcam-video-filter.patch

```diff
@@ -0,0 +1,508 @@
+ OBJS-$(CONFIG_WEAVE_FILTER)                  += vf_weave.o
+ OBJS-$(CONFIG_XBR_FILTER)                    += vf_xbr.o
++OBJS-$(CONFIG_XCAM_FILTER)                   += vf_xcam.o
+ OBJS-$(CONFIG_XFADE_FILTER)                  += vf_xfade.o
+ OBJS-$(CONFIG_XFADE_OPENCL_FILTER)           += vf_xfade_opencl.o opencl.o opencl/xfade.o
+ OBJS-$(CONFIG_XMEDIAN_FILTER)                += vf_xmedian.o framesync.o
+diff --git a/libavfilter/allfilters.c b/libavfilter/allfilters.c
+index 8a149f90a9..80fbfb1e5d 100644
+--- a/libavfilter/allfilters.c
++++ b/libavfilter/allfilters.c
+@@ -435,6 +435,7 @@ extern AVFilter ff_vf_w3fdif;
+ extern AVFilter ff_vf_waveform;
+ extern AVFilter ff_vf_weave;
+ extern AVFilter ff_vf_xbr;
++extern AVFilter ff_vf_xcam;
+ extern AVFilter ff_vf_xfade;
+ extern AVFilter ff_vf_xfade_opencl;
+ extern AVFilter ff_vf_xmedian;
+diff --git a/libavfilter/vf_xcam.c b/libavfilter/vf_xcam.c
+new file mode 100644
+index 0000000000..c9789ee02a
+--- /dev/null
++++ b/libavfilter/vf_xcam.c
+@@ -0,0 +1,317 @@
++/*
++ * Copyright (c) 2020 Intel Corporation, all rights reserved.
++ *
++ * This file is part of FFmpeg.
++ *
++ * FFmpeg is free software; you can redistribute it and/or
++ * modify it under the terms of the GNU Lesser General Public
++ * License as published by the Free Software Foundation; either
++ * version 2.1 of the License, or (at your option) any later version.
++ *
++ * FFmpeg is distributed in the hope that it will be useful,
++ * but WITHOUT ANY WARRANTY; without even the implied warranty of
++ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
++ * Lesser General Public License for more details.
++ *
++ * You should have received a copy of the GNU Lesser General Public
++ * License along with FFmpeg; if not, write to the Free Software
++ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
++ */
++
++/**
++ * @file
++ * libXCam wrapper functions
++ */
++
++#include <xcam/capi/xcam_handle.h>
++#include "libavutil/avstring.h"
++#include "libavutil/opt.h"
++#include "framesync.h"
++#include "internal.h"
++
++typedef struct XCamVideoFilterBuf {
++    XCamVideoBuffer buf;
++    AVFrame *frame;
++} XCamVideoFilterBuf;
++
++typedef struct XCAMContext {
++    const AVClass *class;
++
++    int nb_inputs;
++    int w;
++    int h;
++    char *fmt;
++    char *name;
++    char *params;
++
++    XCamHandle *handle;
++    uint32_t v4l2_fmt;
++
++    XCamVideoFilterBuf **inbufs;
++    XCamVideoFilterBuf *outbuf;
++
++    FFFrameSync fs;
++} XCAMContext;
++
++static uint8_t *xcambuf_map(XCamVideoBuffer *buf)
++{
++    XCamVideoFilterBuf *avfilter_buf = (XCamVideoFilterBuf *)(buf);
++    return avfilter_buf->frame->data[0];
++}
++
++static void xcambuf_unmap(XCamVideoBuffer *buf)
++{
++    return;
++}
++
++static uint32_t avfmt_to_v4l2fmt(int avfmt)
++{
++    if (avfmt == AV_PIX_FMT_YUV420P)
++        return V4L2_PIX_FMT_YUV420;
++    return V4L2_PIX_FMT_NV12;
```

Comment:
what happens if avfmt is not nv12

Reply:
xcam video filter only supports yuv420 and nv12 formats, and the formats have been limited to [yuv420, nv12] in xcam_query_formats().